dashboard/lib/pd/foorm/rollup_creator.rb

```diff
@@ -0,0 +1,116 @@
+# Create rollup for a set of Foorm responses
+# Rollup will only work if question uses numbers for response options (otherwise question will be ignored)
+# Rollup will include average response for each question, and if question is part of a matrix, average for
+# all responses in the matrix. Currently it only returns matrix averages
+module Pd::Foorm
+  class RollupCreator
+    include Constants
+    extend Helper
+
+    # Parameters:
+    #   summarized_answers: output of WorkshopSummarizer.summarize_answers_by_survey
+    #   question_details: output of RollupHelper.get_question_details_for_rollup
+    # Calculate average responses for each question in question_details for the set of responses in summarized_answers
+    # See get_averaged_rollup for response format
+    def self.calculate_averaged_rollup(summarized_answers, question_details)
+      intermediate_rollup = get_intermediate_rollup(summarized_answers, question_details)
+      get_averaged_rollup(intermediate_rollup, question_details)
+    end
+
+    # Calculates average for each question in intermediate rollup and returns rollup in format below.
+    # {
+    #   response_count: 5
+    #   averages: {
+    #     question_id_1: {
+    #       average: 3.45,
+    #       rows: {
+    #         row_id_1: 2.5,
+    #         row_id_2: 5.6
+    #       }
+    #     },...
+    #   }
+    # }
+    def self.get_averaged_rollup(intermediate_rollup, question_details)
+      rollup = {response_count: intermediate_rollup[:response_count], averages: {}}
+      intermediate_rollup[:questions].each do |question, answers|
+        case question_details[question][:type]
+        # TODO: add other answer types
+        when ANSWER_MATRIX
+          rollup[:averages][question] = {}
+          averages = {}
+          overall_sum = 0
+          overall_count = 0
+          answers.each do |matrix_question, matrix_answer|
+            next unless matrix_answer[:count] > 0
```

Comment:
Are we getting the right denominator if we are skipping missing answers?

Reply:
yes, this is calculating the average across all the rows in the matrix, so it is just the overall sum/overall count. If count is 0, sum will also be 0 so no need to add to overall calculation

Reply:
It might be worth noting that for optional questions the denominator is the number of submissions for that answer, not the total number of submissions for the form.